docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,331 @@
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeHolding` (previously `InitializeAccount`)
+* `Revoke`
+* `SetAuthority`
+* `CloseHolding` (previously `CloseAccount`)
+* `FreezeHolding` (previously `FreezeAccount`)
+* `ThawHolding` (previously `ThawAccount`)
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeHolding2` (previously `InitializeAccount2`)
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+* `InitializeMultisig` (use a more general multisig)
+
+New instructions required:
+
+* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+The `amount` field in an SPL token holding should contain a balance in order
+to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.
+
+For i-tokens, `amount` will be in terms of tokens, and not shares, so that UIs
+can properly display amounts moved.
+
+### Token Program Registry
+
+We need the [token-list](https://github.com/solana-labs/token-list) to include
+vetted SPL token-conforming programs and group known mints by their program id.
+
+The token list will publish a file of known program ids at
```

Comment:
This will be the crux of the Ledger app implementation.  Instruction parsing is currently gated on program id and flash storage is limited

Reply:
As MVP, if the Ledger app doesn't recognize the program id (Tokenkeg) then could we display the program id to the user as well for confirmation?

Reply:
Currently it's just going to bail on an unrecognized instruction and the user would have to blind sign.  Without a known program ID there's no way to route to the appropriate instruction parser, so there's nothing else to display the program_id with anyway.  We'll fail similarly on any new instructions in the alternate implementation

I've been debating trying to add a fallback mode that displays instruction info one at a time instead of collecting all of the info up front and coalescing them into a nice display for the whole transaction. I'm not entirely sure that will fit on the device alongside the current implementation though, so it may be one or the other.  I'm also somewhat dubious of this route in that the number of fields the user might be presented with could defeat the purpose of visual confirmation through monotony or making them race the blockhash expiration

Another option would be to move the SPL Token features into a separate Leger app where we can make more assumptions about what it's getting and/or make it easily extensible so projects can just fork it like KIN did the current app.  I'm not 100% sure this is viable though as LedgerHQ may disallow multiple apps using the same derivation path, which would effectively silo them.  I'll ask upstream

Reply:
@t-nelson how does `7320a19` look per our Ledger discussions?  I unfortunately couldn't remember what we decided on the associated token account, so I left a TODO there -- sorry about that!